### PR TITLE
[SPARK-18700][SQL] Add StripedLock for each table's relation in cache

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

As the scenario describe in [SPARK-18700](https://issues.apache.org/jira/browse/SPARK-18700), when cachedDataSourceTables invalided, the coming few queries will fetch all FileStatus in listLeafFiles function. In the condition of table has many partitions, these jobs will occupy much memory of driver finally may cause driver OOM.

In this patch, add StripedLock for each table's relation in cache not for the whole cachedDataSourceTables, each table's load cache operation protected by it.

## How was this patch tested?

Add a multi-thread access table test in `PartitionedTablePerfStatsSuite` and check it only loading once using metrics in `HiveCatalogMetrics`

### CODE REVIEW

**Comment by xuanyuanking** [comment]
@rxin @liancheng

**Comment by rxin** [comment]
cc @ericl  can you take a look at this?

**Comment by ericl** [comment]
Isn't it sufficient to lock around the `catalog.filterPartitions(Nil)`? Why do we need reader locks?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 1]
```diff
@@ -53,6 +56,10 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
       tableIdent.table.toLowerCase)
   }
 
+  /** ReadWriteLock for each tables, protect the read and write cached */
+  protected[hive] val tableLockMap =
+    new ConcurrentHashMap[QualifiedTableName, ReentrantReadWriteLock]
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/Striped.html is usually good enough

Comment (xuanyuanking) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Yes, I considered the Striped.lazyWeakReadWriteLock here but I need deal with invalidateAllCache(), it need all <K, V>, so I finally new a HashMap self.

**Comment by xuanyuanking** [comment]
@ericl Thanks for your review.

> Is it sufficient to lock around the catalog.filterPartitions(Nil)?
Yes, this patch port from 1.6.2 and I missed the diff here. Fixed in next patch.

>  Why do we need reader locks?
Write or Invalid the table cache operation fewer than read it. Reader waiting when there is same table writing cache.

**Comment by ericl** [comment]
I guess the large number of lock sites is confusing me. We only want to prevent concurrent instantiation of a single table, so shouldn't you only need 1 lock for that site?

Also, we should have a unit test that tries to concurrently read from a table from many threads, and verifies via the catalog metrics that it is only loaded once (see `TablePerfStatsSuite` for how to access the metrics).

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 2]
```diff
@@ -105,7 +105,8 @@ private[sql] class HiveSessionCatalog(
   // For testing only
   private[hive] def getCachedDataSourceTable(table: TableIdentifier): LogicalPlan = {
     val key = metastoreCatalog.getQualifiedTableName(table)
-    metastoreCatalog.cachedDataSourceTables.getIfPresent(key)
+    metastoreCatalog.readLock(key,
+      metastoreCatalog.cachedDataSourceTables.getIfPresent(key))
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: Why a read lock here?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 3]
```diff
@@ -95,7 +95,7 @@ private[sql] class HiveSessionCatalog(
   }
 
   def invalidateCache(): Unit = {
-    metastoreCatalog.cachedDataSourceTables.invalidateAll()
+    metastoreCatalog.invalidateAllCache()
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: Why this change?

**Comment by xuanyuanking** [comment]
hi @ericl 
This commit do the 3 things below, thanks for your check:
1. Delete the unnecessary lock use and simplify the lock operation
2. Add UT test in `PartitionedTablePerfStatsSuite`
3. Add cache hit metrics in `HiveCatalogMetrics`
Also change the description of this PR.

File: core/src/main/scala/org/apache/spark/metrics/source/StaticSources.scala [hunk 4]
```diff
@@ -97,6 +97,12 @@ object HiveCatalogMetrics extends Source {
     MetricRegistry.name("parallelListingJobCount"))
 
   /**
+   * Tracks the total number of cachedDataSourceTables hits.
+   */
+  val METRIC_DATASOUCE_TABLE_CACHE_HITS = metricRegistry.counter(
+    MetricRegistry.name("dataSourceTableCacheHits"))
```
Comment (ericl) [inline core/src/main/scala/org/apache/spark/metrics/source/StaticSources.scala]: Could we use one of the other metrics, rather than add a new one?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 5]
```diff
@@ -33,6 +35,7 @@ import org.apache.spark.sql.hive.orc.OrcFileFormat
 import org.apache.spark.sql.types._
 
 
+
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: nit: extra newline

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 6]
```diff
@@ -53,6 +56,18 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
       tableIdent.table.toLowerCase)
   }
 
+  /** ReadWriteLock for each tables, protect the read and write cached */
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Could you update this comment to say that the reason we lock is to prevent concurrent table instantiation?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 7]
```diff
@@ -53,6 +56,18 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
...
+  private val tableLockStripes = Striped.lazyWeakLock(10)
+
+  /** Acquires a lock on the table cache for the duration of `f`. */
+  private def cacheLock[A](tableName: QualifiedTableName, f: => A): A = {
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: withTableCreationLock

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala [hunk 8]
```diff
@@ -352,4 +353,28 @@ class PartitionedTablePerfStatsSuite
...
+              spark.sql("select * from test where partCol1 = 999").count()
+            }
+          }
+          executorPool.execute(runnable)
+          None
+        })
+        executorPool.shutdown()
+        executorPool.awaitTermination(30, TimeUnit.SECONDS)
+        // check the cache hit, the cache only load once
+        assert(HiveCatalogMetrics.METRIC_DATASOUCE_TABLE_CACHE_HITS.getCount() == 9)
```
Comment (ericl) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: Does this test fail without the lock?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala [hunk 9]
```diff
@@ -352,4 +353,28 @@ class PartitionedTablePerfStatsSuite
       }
     }
   }
+
+  test("SPARK-18700: add lock for each table's realation in cache") {
```
Comment (ericl) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: "table loaded only once even when resolved concurrently"

**Review by ericl — CHANGES_REQUESTED** [review]

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 10]
```diff
@@ -53,6 +56,18 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
       tableIdent.table.toLowerCase)
   }
 
+  /** ReadWriteLock for each tables, protect the read and write cached */
+  private val tableLockStripes = Striped.lazyWeakLock(10)
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Hm, may as well make it 100 if it's a lazy weak locks.

Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: nit: `tableCreationLocks`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 5 above)
Comment (xuanyuanking) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: fix done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 6 above)
Comment (xuanyuanking) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: update done, and more comments at HiveMetastoreCatalog.scala+226

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 10 above)
Comment (xuanyuanking) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: fix done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 7 above)
Comment (xuanyuanking) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: update done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala (hunk 9 above)
Comment (xuanyuanking) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: update done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala (hunk 8 above)
Comment (xuanyuanking) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: It may failed sometimes without lock, but when I add `Thread.sleep(1000)` before `cachedDataSourceTables.put(tableIdentifier, created)` in HiveMetastoreCatalog.scala +265 to make the build table relation slow.It will failed every time. How can I do this hook in UT? Or how to make the cache build operation slow without really make a big table?  : )

Comment (ericl) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: Yeah that's fine, as long as it fails some fraction of the time it will eventually show up as a flaky test.

File: core/src/main/scala/org/apache/spark/metrics/source/StaticSources.scala [hunk 11]
```diff
@@ -105,6 +111,7 @@ object HiveCatalogMetrics extends Source {
     METRIC_FILE_CACHE_HITS.dec(METRIC_FILE_CACHE_HITS.getCount())
     METRIC_HIVE_CLIENT_CALLS.dec(METRIC_HIVE_CLIENT_CALLS.getCount())
     METRIC_PARALLEL_LISTING_JOB_COUNT.dec(METRIC_PARALLEL_LISTING_JOB_COUNT.getCount())
+    METRIC_DATASOUCE_TABLE_CACHE_HITS.dec(METRIC_DATASOUCE_TABLE_CACHE_HITS.getCount())
```
Comment (ericl) [inline core/src/main/scala/org/apache/spark/metrics/source/StaticSources.scala]: s/souce/source

File: core/src/main/scala/org/apache/spark/metrics/source/StaticSources.scala (hunk 4 above)
Comment (ericl) [inline core/src/main/scala/org/apache/spark/metrics/source/StaticSources.scala]: That's kind of odd, I'd expect the duplicate table building to cause more file accesses or at least cache hits since we are scanning the filesystem multiple times. Is that not the case?

Comment (xuanyuanking) [inline core/src/main/scala/org/apache/spark/metrics/source/StaticSources.scala]: I may find the reason of the odd scenario, please check my conclusion:
In 2.0 add a new config lazyPruningEnabled, it's default value is true and while multi-thread do the building same time, it will not do listLeafFile.
So I set the `HIVE_MANAGE_FILESOURCE_PARTITIONS`=false and set the partition larger than `PARALLEL_PARTITION_DISCOVERY_THRESHOLD`(this will cause METRIC_PARALLEL_LISTING_JOB_COUNT +1), test results list below:
without lock:
```
0
0
550   (50 file * 11, 1 is cache.load() and the other 10 is 10 threads)
90
11      (also 1 * 11)
1000
```
and with lock:
```
9
0
100  (2 * 50, 1 is cache.load() and the other 1 is the first threads)
54
2     (also 1 * 2)
550
```
so I can delete the added `dataSourceTableCacheHits` metrics and use `parallelListingJobCount` and `filesDiscovered` instead. Do this in next patch.

File: core/src/main/scala/org/apache/spark/metrics/source/StaticSources.scala (hunk 11 above)
Comment (xuanyuanking) [inline core/src/main/scala/org/apache/spark/metrics/source/StaticSources.scala]: e...sorry, this new added metric will delete next patch like before comment

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 12]
```diff
@@ -209,72 +221,79 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
...
-        val fileCatalog = {
-          val catalog = new CatalogFileIndex(
-            sparkSession, metastoreRelation.catalogTable, sizeInBytes)
-          if (lazyPruningEnabled) {
-            catalog
-          } else {
-            catalog.filterPartitions(Nil)  // materialize all the partitions in memory
+      // Here we should protect all relation get and create operation with lock while big
+      // table's CatalogFileIndex will take some time, only lock cachedDataSourceTables.put
+      // will still cause driver memory waste. More detail see SPARK-18700.
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: nit: probably don't need this comment

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 13]
```diff
@@ -53,6 +53,18 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
       tableIdent.table.toLowerCase)
   }
 
+  /** Locks for preventing driver mem waste when concurrent table instantiation */
+  private val tableCreationLocks = Striped.lazyWeakLock(100)
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: nit: "These locks guard against multiple attempts to instantiate a table, which wastes memory."

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala [hunk 14]
```diff
@@ -352,4 +353,34 @@ class PartitionedTablePerfStatsSuite
...
+            }
+            executorPool.execute(runnable)
+            None
+          })
+          executorPool.shutdown()
+          executorPool.awaitTermination(30, TimeUnit.SECONDS)
+          // check the cache hit, we use the metric of METRIC_FILES_DISCOVERED and
+          // METRIC_PARALLEL_LISTING_JOB_COUNT to check this, while the lock take effect,
+          // only one thread can really do the build, so the listing job count is 2, the other
+          // one is cahce.load func. Also METRIC_FILES_DISCOVERED is $partition_num * 2
```
Comment (ericl) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: s/cahce/cache

**Review by ericl — APPROVED** [review]
Thanks for making these changes, this lgtm with some nits.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 13 above)
Comment (xuanyuanking) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: fix done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 12 above)
Comment (xuanyuanking) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: done, delete it

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala (hunk 14 above)
Comment (xuanyuanking) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: ……sorry, fix done

**Comment by xuanyuanking** [comment]
Thanks for ericl's review!

**Comment by ericl** [comment]
cc @rxin

**Comment by xuanyuanking** [comment]
cc @rxin thanks for check.

**Comment by hvanhovell** [comment]
retest this please

**Comment by hvanhovell** [comment]
I am merging this one after a successful test run. Ping me if you object.

**Comment by rxin** [comment]
Go for it.

**Comment by hvanhovell** [comment]
Merging to master/2.1. @xuanyuanking can you open a backport for 2.0, if we also need to merge this to that branche?

**Comment by xuanyuanking** [comment]
@hvanhovell Sure, I open a new BACKPORT-2.0.
There's a little diff in branch-2.0, the ut test of this patch based on the `HiveCatalogMetrics` which not added in 2.0, so I added the patch need metric. Thanks for check.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala [hunk 15]
```diff
@@ -352,4 +353,34 @@ class PartitionedTablePerfStatsSuite
...
+            }
+            executorPool.execute(runnable)
+            None
+          })
+          executorPool.shutdown()
+          executorPool.awaitTermination(30, TimeUnit.SECONDS)
+          // check the cache hit, we use the metric of METRIC_FILES_DISCOVERED and
+          // METRIC_PARALLEL_LISTING_JOB_COUNT to check this, while the lock take effect,
+          // only one thread can really do the build, so the listing job count is 2, the other
+          // one is cache.load func. Also METRIC_FILES_DISCOVERED is $partition_num * 2
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: This comment is wrong. The extra counts are from the `DataFrameWriter`'s save() API.

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: Working on a fix to avoid the useless filesystem scan caused by the save() API.

Comment (xuanyuanking) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala]: @gatorsmile Xiao fixed this in https://github.com/apache/spark/pull/16481
